control-plane/connect-inject/controllers/endpointsv2/endpoints_controller.go

```diff
@@ -0,0 +1,351 @@
+// prefix. Currently, this only applies to ReplicaSets.
+//
+// We have to fetch the owner and check its type because pod names cannot be disambiguated from pod owner names due to
+// the `-` delimiter and unique ID parts also being valid name components.
+//
+// If the pod owner does not have a unique name, the empty string is returned.
+func getOwnerPrefixFromPod(pod *corev1.Pod) string {
+	for _, ref := range pod.OwnerReferences {
+		if ref.Kind == "ReplicaSet" {
+			return ref.Name
+		}
+	}
+	return ""
+}
+
+// registerService creates a Consul service registration from the provided Kuberetes service and endpoint information.
+func (r *Controller) registerService(ctx context.Context, resourceClient pbresource.ResourceServiceClient, service corev1.Service, podPrefixes, podExactNames map[string]any) error {
+	serviceResource := r.getServiceResource(
+		&pbcatalog.Service{
+			Workloads:  getWorkloadSelector(podPrefixes, podExactNames),
+			Ports:      getServicePorts(service),
+			VirtualIps: r.getServiceVIPs(service),
+		},
+		service.Name, // Consul and Kubernetes service name will always match
+		r.getConsulNamespace(service.Namespace),
+		r.getConsulPartition(),
+		getServiceMeta(service),
+	)
+
+	r.Log.Info("registering service with Consul", getLogFieldsForResource(serviceResource.Id)...)
+	//TODO(zalimeni) Read from Consul or some other state to determine whether a write is needed.
+	// For now, we blindly write state on each reconcile to simplify implementation.
+	_, err := resourceClient.Write(ctx, &pbresource.WriteRequest{Resource: serviceResource})
+	if err != nil {
+		r.Log.Error(err, "failed to register service", getLogFieldsForResource(serviceResource.Id)...)
+		return err
+	}
+
+	return nil
+}
+
+// getServiceResource converts the given Consul service and metadata as a Consul resource API record.
+func (r *Controller) getServiceResource(svc *pbcatalog.Service, name, namespace, partition string, meta map[string]string) *pbresource.Resource {
+	return &pbresource.Resource{
+		Id:       getServiceID(name, namespace, partition),
+		Data:     common.ToProtoAny(svc),
+		Metadata: meta,
+	}
+}
+
+func getServiceID(name, namespace, partition string) *pbresource.ID {
+	return &pbresource.ID{
+		Name: name,
+		Type: &pbresource.Type{
+			Group:        "catalog",
+			GroupVersion: "v1alpha1",
+			Kind:         "Service",
+		},
+		Tenancy: &pbresource.Tenancy{
+			Partition: partition,
+			Namespace: namespace,
+		},
+	}
+}
+
+// getServicePorts converts Kubernetes Service ports data into Consul service ports.
+func getServicePorts(service corev1.Service) []*pbcatalog.ServicePort {
+	ports := make([]*pbcatalog.ServicePort, 0, len(service.Spec.Ports)+1)
+
+	for _, p := range service.Spec.Ports {
+		ports = append(ports, &pbcatalog.ServicePort{
+			VirtualPort: uint32(p.Port),
+			TargetPort:  p.TargetPort.String(),
+			Protocol:    common.GetPortProtocol(p.AppProtocol),
```

Comment:
I'd appreciate a spot check on this mapping JIC I'm missing some nuance - cc @ishustava in case you have input

Reply:
I don't think `targetPort` will work in the case it is a number, but the pod defines the service with a name. We will need to look up the corresponding port name in the pod.  I'm open to changing how this works on the Pod side but not sure how you get around the name being optional and `targetPort` working on either.

Reply:
> I don't think targetPort will work in the case it is a number, but the pod defines the service with a name

(_This might be wrong or obvious, feel free to correct me_) It seems to me from the k8s [docs](https://kubernetes.io/docs/concepts/services-networking/service/#defining-a-service) that there's an inherent tradeoff that applies generally: you need to pick whether to use a name consistently across service and pods (which enables mixed pod port numbers under one name), or you need to pick a consistent pod port number and use that in the service. Either way, it seems selected pods need to adhere to the service's chosen int-or-string value for `targetPort`.

Our docs for Consul `Service.TargetPort` say `target_port is the name of the workload port`, so I took that to mean we defer totally to the value used by the k8s service.

It seems wrong to me to reach into the pods to infer anything beyond the workload selector, since the `targetPort` mapping isn't really a `consul-k8s`-specific limitation... But I might be missing a reason we'd want to do that.

> I'm open to changing how this works on the Pod side

What did you have in mind here?

Reply:
> It seems wrong to me to reach into the pods to infer anything beyond the workload selector, since the targetPort mapping isn't really a consul-k8s-specific limitation

Yeah, I don't like the idea of reaching in either. The pods may be slightly different, and the names don't align, making the lookup challenging. But I think I disagree that there isn't something K8s-specific here. We generate a Consul port name per port, even if one wasn't defined in the Kube spec, and Kubernetes doesn't necessarily require you to use any name at all, so there is a mismatch of strong (consul) vs. weak (k8s) references. I think I had the opposite reaction that since the VM case is straightforward, we'd have to "fix it" in consul-k8s.

>> I'm open to changing how this works on the Pod side
No specific ideas here. I'm just trying not to look like I'm suggesting a singular solution.

Reply:
Started 🧵 [here](https://hashicorp.slack.com/archives/C04MG2UT3PA/p1694005024255509) to figure out how we want this to work